File: src/tps/tp1/ex11_PalindromoRecursivoC/palindromoRecursivo.c
/**
 * Bernardo Vieira Rocha - 853733
 * Ciencia da Computacao PUC-MINAS
 * TP01_Q11 - Palindromo Recursivo (palindromoRecursivoC)
*/

// TODO: Porque o codigo ideal pensando no primeiro caso nao funciona mais?

#include <stdio.h>
#include <string.h>
#include <stdbool.h>

/**
 * Funcao que verifica se uma palavra eh palindromo
 * @param palavra Palavra a ser verificada
 * @param direita Indice da direita
 * @param esquerda Indice da esquerda
 * @return true se a palavra eh palindromo, false caso contrario
 */
bool palindromo(char palavra[], int direita, int esquerda) {
    if (direita >= esquerda) {
        return true;
    } else if (palavra[direita] != palavra[esquerda]) {
        return false;
    } else {
        return palindromo(palavra, direita + 1, esquerda - 1);
    }
}

int main() {
    char palavra[1000];

    while (true) {
        // Le a palavra ate a quebra de linha 
        scanf(" %[^\n]", palavra);

        // Checa se o input eh FIM
        if (strcmp(palavra, "FIM") == 0) {
            break;
        } else {
            int len = strlen(palavra);
            if (palindromo(palavra, 0, len - 1)) {
                printf("SIM\n");
            } else {
                printf("NAO\n");
            }
        }
    }
    return 0;
}

File: src/tps/tp1/ex10_palindromoRecursivoJava/palindromoRecursivo.java
/*
 * Bernardo Vieira Rocha - 853733
 * Ciencia da Computacao PUC-MINAS
 * TP01_Q10- Palindromo RECURSIVO (palindromoRecursivoJava)
 */

/*
 * Classe que verifica se uma string é um palindromo de forma recursiva.
 */
@SuppressWarnings("WrongPackageStatement")
public class palindromoRecursivo {

    /*
     * Metodo auxiliar que inicializa a verificacao do palindromo.
     * @param input - String a ser verificada
     * @return boolean - Resposta da verificação
     */
    public static boolean isPalindromo(String input) {
        return isPalindromo(input, 0);
    }

    /*
    *  Verifica de maneira recursiva se a string eh palindroma
     * @param input - String a ser verificada
     * @param count - Contador de posicoes da string
     * @return boolen - Resposta da verificacao
    * */
    public static boolean isPalindromo(String input, int count) {
        if (count >= input.length() / 2) {
            return true;
        }
        if (input.charAt(count) != input.charAt(input.length() - count - 1)) {
            return false;
        }
        return isPalindromo(input, count + 1);
    }

    /*
     * Funcao que checa se a string eh igual a FIM
     * @param input - String a ser verificada
     * @return true se a string for igual e false caso contrario
    * */
    public static boolean isFIM(String input){
        return (input.length() == 3 && input.charAt(0) == 'F' && input.charAt(1) == 'I' && input.charAt(2) == 'M');
    }

    public static void main(String[] args) {
        String s;
        MyIO.setCharset("UTF-8");
        while (!isFIM((s = MyIO.readLine())) ){
            System.out.println(isPalindromo(s, 0) ? "SIM" : "NAO");
        }
    }
}
